homeassistant/components/lcn/websocket.py

```diff
@@ -0,0 +1,419 @@
+)
+@websocket_api.async_response
+async def websocket_get_entity_configs(
+    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+) -> None:
+    """Get entities configs."""
+    config_entry = hass.config_entries.async_get_entry(msg["host_id"])
+    if config_entry is None:
+        return
+
+    if CONF_ADDRESS in msg:
+        entity_configs = [
+            entity_config
+            for entity_config in config_entry.data[CONF_ENTITIES]
+            if tuple(entity_config[CONF_ADDRESS]) == msg[CONF_ADDRESS]
+        ]
+    else:
+        entity_configs = config_entry.data[CONF_ENTITIES]
+
+    connection.send_result(msg["id"], entity_configs)
+
+
+@websocket_api.require_admin
+@websocket_api.websocket_command(
+    {vol.Required("type"): "lcn/devices/scan", vol.Required("host_id"): cv.string}
+)
+@websocket_api.async_response
+async def websocket_scan_devices(
+    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+) -> None:
+    """Scan for new devices."""
+    config_entry = hass.config_entries.async_get_entry(msg["host_id"])
+    if config_entry is None:
+        return
+
+    host_connection = hass.data[DOMAIN][config_entry.entry_id][CONNECTION]
+    await host_connection.scan_modules()
+
+    for device_connection in host_connection.address_conns.values():
+        if not device_connection.is_group:
+            await async_create_or_update_device_in_config_entry(
+                hass, device_connection, config_entry
+            )
+
+    # create/update devices in device registry
+    register_lcn_address_devices(hass, config_entry)
+
+    connection.send_result(msg["id"], config_entry.data[CONF_DEVICES])
+
+
+@websocket_api.require_admin
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "lcn/devices/add",
+        vol.Required("host_id"): cv.string,
+        vol.Required(CONF_ADDRESS): ADDRESS_SCHEMA,
+        vol.Required(CONF_NAME): cv.string,
+    }
+)
+@websocket_api.async_response
+async def websocket_add_device(
+    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+) -> None:
+    """Add a device."""
+    config_entry = hass.config_entries.async_get_entry(msg["host_id"])
+    if config_entry is None:
+        return
+
+    if get_device_config(msg[CONF_ADDRESS], config_entry):
+        connection.send_result(
+            msg["id"], False
+        )  # device_config already in config_entry
+        return
+
+    device_config = {
+        CONF_ADDRESS: msg[CONF_ADDRESS],
+        CONF_NAME: "",
+        CONF_HARDWARE_SERIAL: -1,
+        CONF_SOFTWARE_SERIAL: -1,
+        CONF_HARDWARE_TYPE: -1,
+    }
+
+    # update device info from LCN
+    device_connection = get_device_connection(hass, msg[CONF_ADDRESS], config_entry)
+    await async_update_device_config(device_connection, device_config)
+
+    # add device_config to config_entry
+    data = deepcopy(dict(config_entry.data))
+    data[CONF_DEVICES].append(device_config)
```

Comment:
The deepcopy seems unnecessary, how about this:
```suggestion
    device_configs = [*config_entry.data[CONF_DEVICES], device_config]
    data = {**config_entry.data, CONF_DEVICES: device_configs}
```

Same comment for the other places where deepcopy is used

Reply:
Changed.